File: Array/C++/1_Two_Sum.cpp
//Unordered Map(Hash Map) (Time: O(N) Space: O(N))
class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        unordered_map<int, int> numDict;
        vector<int> ans;
        
        for(int i = 0; i < nums.size(); i++) {
            int compliment = target - nums[i];
            if(numDict.find(compliment) != numDict.end()) {
                // ans.push_back(i);
                // ans.push_back(numDict.at(compliment));
                // return ans;
                return {numDict[compliment], i};
            }
            numDict[nums[i]] = i;
        }
        // return ans;
        return {};
    }
};

//Brute Force (Time: O(N^2) Space: O(1))
class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        // vector<int> ans;
        
        for(int i = 0; i < nums.size(); i++) {
            for(int j = i + 1; j < nums.size(); j++) {
                if(nums[i] + nums[j] == target) {
                    // ans.push_back(i);
                    // ans.push_back(j);
                    // return ans;
                    return {i, j};
                }
            }
        }
        // return ans;
        return {};
    }
};

File: Array/Python3/1_Two_Sum.py
# Nested loop solution (Brute Force)
# O(n^2) Time Complexity
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        n = len(nums)    # The array size
        # The outer loop loops through 0 to (n - 1) indices and
        # the inner loop loops through (i + 1) to n indices respectively
        for i in range(n - 1):
            for j in range(i + 1, n):
                if(nums[i] + nums[j] == target):
                    return [i, j]

# Hashmap solution (One-pass Hash Table)
# O(n) Time Complexity
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        n = len(nums)   # The array size
        numMap = {}     # Hash table

        # Loop through the array and
        # check the complement of target exists
        for i in range(n):
            complement = target - nums[i]    # Compute the complement of nums[i]
          
            # Check the complement exists in the hash table
            # If it does, return the indices
            if(complement in numMap):
                return [numMap[complement], i]

            # Add the ith element in the hash table
            numMap[nums[i]] = i
